group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1043,4 +1221,1331 @@ public void onNewMetadataImage(MetadataImage newImage, MetadataDelta delta) {
+            default:
+                throw new IllegalStateException("Unknown group state: " + group.stateAsString());
+        }
+    }
+
+    /**
+     * Update a static member and rebalance.
+     *
+     * @param context         The request context.
+     * @param request         The join group request.
+     * @param group           The group of the static member.
+     * @param oldMemberId     The existing static member id.
+     * @param newMemberId     The new joining static member id.
+     * @param joinReason      The client reason for the join request.
+     * @param responseFuture  The response future to complete.
+     *
+     * @return The coordinator result that will be appended to the log.
+     */
+    private CoordinatorResult<Void, Record> updateStaticMemberAndRebalance(
+        RequestContext context,
+        JoinGroupRequestData request,
+        GenericGroup group,
+        String oldMemberId,
+        String newMemberId,
+        String joinReason,
+        CompletableFuture<JoinGroupResponseData> responseFuture
+    ) {
+        String currentLeader = group.leaderOrNull();
+        GenericGroupMember newMember = group.replaceStaticMember(request.groupInstanceId(), oldMemberId, newMemberId);
+
+        // Heartbeat of old member id will expire without effect since the group no longer contains that member id.
+        // New heartbeat shall be scheduled with new member id.
+        rescheduleGenericGroupMemberHeartbeat(group, newMember);
+
+        int oldRebalanceTimeoutMs = newMember.rebalanceTimeoutMs();
+        int oldSessionTimeoutMs = newMember.sessionTimeoutMs();
+        JoinGroupRequestProtocolCollection oldProtocols = newMember.supportedProtocols();
+
+        group.updateMember(
+            newMember,
+            request.protocols(),
+            request.rebalanceTimeoutMs(),
+            request.sessionTimeoutMs(),
+            responseFuture
+        );
+
+        if (group.isInState(STABLE)) {
+            // Check if group's selected protocol of next generation will change, if not, simply store group to persist
+            // the updated static member, if yes, rebalance should be triggered to keep the group's assignment
+            // and selected protocol consistent
+            String groupInstanceId = request.groupInstanceId();
+            String selectedProtocolForNextGeneration = group.selectProtocol();
+            if (group.protocolName().orElse("").equals(selectedProtocolForNextGeneration)) {
+                log.info("Static member which joins during Stable stage and doesn't affect " +
+                    "the selected protocol will not trigger a rebalance.");
+
+                CompletableFuture<Void> appendFuture = new CompletableFuture<>();
+                appendFuture.whenComplete((__, t) -> {
+                    if (t != null) {
+                        log.warn("Failed to persist metadata for group {}: {}", group.groupId(), t.getMessage());
```

Comment:
nit: Could we expand this error a little? It would be great if it could capture that it failed to update the metadata for a static member, etc.